### ISSUE TITLE
Refactor to support plugin like babel

### ISSUE THREAD

**Issue by sibelius** [lucasbento/create-graphql#59]

**Comment by sibelius** [comment]
First draft of Plugins

## Read Database Schema Plugins
### Motivation
A read schema plugin should read a schema (collection/table definition) and return an object describing it.

### Examples
- create-graphql-mongoose
- create-graphql-mysql
- create-graphql-postgres
- create-graphql-rethinkdb

## Templates Plugins
### Motivation
CreateGraphql should output a different file for a different option provided, for instance: output a raw mutation or a relay mutation

### Examples
- create-graphql-type-raw
- create-graphql-type-relay
- create-graphql-mutation-raw
- create-graphql-mutation-relay

we need to define inputs and outputs of these plugins

**Comment by lucasbento** [comment]
I can imagine it to handle schema from multiple types of databases but how would that be for different templates?

**Comment by lucasbento** [comment]
Also, how do you plan on starting, should #43 wait for it?

Turning the `mongoose` schema reading into a plugin should be really straightforward.

We should also have a `plugins` property on `.graphqlrc`.

**Comment by sibelius** [comment]
A template plugin could have an option parameter, like -t for template

**Comment by lucasbento** [comment]
> A template plugin could have an option parameter, like -t for template

@sibelius: what do you mean?

I meant, how will we implement a template plugin? Will we prepare all the variables, pass it to the plugin and it will return the files?

**Comment by sibelius** [comment]
Yep, I think so, we need to define the inputs and outputs of each plugin

**Comment by lucasbento** [comment]
So, I started working on this, I thought the following as a blueprint for a `create-graphql-plugin`:

```js
export default {
  pre: (arguments) => {
    const {
      config, // The content of `.graphqlrc`, will be the default if it doesn't exist
    } = arguments;

    console.log('this is a pre hook function');

    return {
      fields, // The fields generated on the plugin
      files, // Array of objects, e.g.: [{ name: 'ExampleAddMutation.js', content: 'Content of file }]
    };
  },
  post: (arguments) => {
    const {
      config, // The content of `.graphqlrc`, will be the default if it doesn't exist,
      fields, // The fields of the schema
    } = arguments;

    console.log('this is a post hook function');

    return {}; // What would this return?
  },
};
```

Would love your opinions on this, @sibelius & @felippepuhle.

Also, the plugins would be installed as a `devDependency` and specified on `.graphqlrc` under `plugins`, e.g.:
```json
{
  "directories": {
    ...
    "connection": "connection",
    "loader": "loader",
    "model": "model",
    ...
  },
  "plugins": [
    "create-graphql-plugin-mongoose",
    "create-graphql-plugin-sequelize",
    ...
  ]
}
```

**Comment by lucasbento** [comment]
Good content on Babel plugins: https://github.com/thejameskyle/babel-handbook/blob/master/translations/en/plugin-handbook.md

**Comment by lucasbento** [comment]
https://github.com/thejameskyle/babel-file
